crates/credential/cargo-credential/src/lib.rs

```diff
@@ -0,0 +1,81 @@
+//! Helper library for writing Cargo credential processes.
+//!
+//! A credential process should have a `struct` that implements the `Credential` trait.
+//! The `main` function should be called with an instance of that struct, such as:
+//!
+//! ```rust,ignore
+//! fn main() {
+//!     cargo_credential::main(MyCredential);
+//! }
+//! ```
+
+pub type Error = Box<dyn std::error::Error>;
+
+pub trait Credential {
+    /// Returns the name of this credential process.
+    fn name(&self) -> &'static str;
+
+    /// Retrieves a token for the given registry.
+    fn get(&self, registry_name: &str, api_url: &str) -> Result<String, Error>;
+
+    /// Stores the given token for the given registry.
+    fn store(&self, registry_name: &str, api_url: &str, token: &str) -> Result<(), Error>;
+
+    /// Removes the token for the given registry.
+    ///
+    /// If the user is not logged in, this should print a message to stderr if
+    /// possible indicating that the user is not currently logged in, and
+    /// return `Ok`.
+    fn erase(&self, registry_name: &str, api_url: &str) -> Result<(), Error>;
+}
+
+/// Runs the credential interaction by processing the command-line and
+/// environment variables.
+pub fn main(credential: impl Credential) {
+    let name = credential.name();
+    if let Err(e) = doit(credential) {
+        eprintln!("{} error: {}", name, e);
+        std::process::exit(1);
+    }
+}
+
+fn env(name: &str) -> Result<String, Error> {
+    std::env::var(name).map_err(|_| format!("environment variable `{}` is not set", name).into())
+}
+
+fn doit(credential: impl Credential) -> Result<(), Error> {
+    let which = std::env::args()
```

Comment:
This behavior of discovering the action may be worth mentioning in the crate docs?